src/core_landice/mode_forward/mpas_li_calving.F

```diff
@@ -1520,56 +1539,71 @@ subroutine apply_calving_velocity(meshPool, geometryPool, velocityPool, domain,
+            do iEdge = 1, nEdgesOnCell(iCell)
+              iNeighbor = cellsOnCell(iEdge, iCell)
+              if ( (((thickness(iNeighbor) == 0.0_RKIND) &
+                 .and. bedTopography(iNeighbor) < config_sea_level)) &
+                 .or. ((li_mask_is_floating_ice(cellMask(iNeighbor)) &
+                 .and. li_mask_is_margin(cellMask(iNeighbor)) &
+                 .and. (.not. li_mask_is_dynamic_ice(cellMask(iNeighbor)))))) then
+                 nEmptyNeighbors = nEmptyNeighbors + 1
+              endif
+            enddo
 
-         if ( nEmptyNeighbors > 0 &
-           .and. li_mask_is_grounded_ice(cellMask(iCell)) &
-           .and. bedTopography(iCell) < config_sea_level &
-           .and. li_mask_is_dynamic_ice(cellMask(iCell)) ) then
+            if ( nEmptyNeighbors > 0 &
+               .and. li_mask_is_grounded_ice(cellMask(iCell)) &
+               .and. bedTopography(iCell) < config_sea_level &
+               .and. li_mask_is_dynamic_ice(cellMask(iCell)) ) then
 
-            groundedMarineMarginMask(iCell) = 1
-         else
-            groundedMarineMarginMask(iCell) = 0
-         endif
-      enddo
+               groundedMarineMarginMask(iCell) = 1
+            else
+               groundedMarineMarginMask(iCell) = 0
+            endif
+         enddo
+      endif
 
 
       ! Init fields for accounting
       calvingThickness(:) = 0.0_RKIND
-      calvedVolumeNonDynCell(:) = 0.0_RKIND
-      calvedVolumeDynCell(:) = 0.0_RKIND
-      uncalvedVolumeNonDynCell(:) = 0.0_RKIND
-      uncalvedVolumeDynCell(:) = 0.0_RKIND
+      meltingThickness(:) = 0.0_RKIND
+      
+      ! Init fields for accounting
+      allocate(ablationThickness(nCells+1))
+      ablationThickness(:) = 0.0_RKIND
+      unablatedVolumeNonDynCell(:) = 0.0_RKIND
+      unablatedVolumeDynCell(:) = 0.0_RKIND
       allocate(cellVolume(nCells+1))
       cellVolume(:) = areaCell(:) * thickness(:)
-      requiredCalvingVolumeNonDynEdge(:) = 0.0_RKIND
-      requiredCalvingVolumeNonDynCell(:) = 0.0_RKIND
-      requiredCalvingVolumeDynEdge(:) = 0.0_RKIND
-      requiredCalvingVolumeDynCell(:) = 0.0_RKIND
-      allocate(thicknessForCalving(nCells+1))
-      thicknessForCalving = 0.0_RKIND
-
-
-
-      ! 1. Calculate calving rate for all non-dynamic cells by working with their ocean-going edges
-      calvingSubtotal1 = 0.0_RKIND
+      allocate(requiredAblationVolumeNonDynEdge(nEdges+1))
+      allocate(requiredAblationVolumeNonDynCell(nCells+1))
+      allocate(requiredAblationVolumeDynEdge(nEdges+1))
+      allocate(requiredAblationVolumeDynCell(nCells+1))
+      requiredAblationVolumeNonDynEdge(:) = 0.0_RKIND
+      requiredAblationVolumeNonDynCell(:) = 0.0_RKIND
+      requiredAblationVolumeDynEdge(:) = 0.0_RKIND
+      requiredAblationVolumeDynCell(:) = 0.0_RKIND
+      allocate(thicknessForAblation(nCells+1))
+      thicknessForAblation = 0.0_RKIND
+      allocate(ablationVelocity(nCells+1))
+
+      if ( applyCalving ) then
```

Comment:
This has been updated. We now pass the desired calving or melting thickness and velocity arrays to li_apply_ablation_velocity, which no longer differentiates between calving and melting. We vertically average melting over the whole ice face before calling li_apply_ablation_velocity, since faceMeltSpeed is only below the waterline. Halo updates are now applied in the calling routines.